ciba-grant-3.2.0.md

shilpa-padgaonkar opened the pull request:
Below are details of the proposed changes in this PR to add CIBA grant flow:

- The supported  OAuth flows list has been extended with the CIBA flow

<img width="580" alt="image" src="https://github.com/OAI/OpenAPI-Specification/assets/77152136/79d1f0e5-bd5f-4f1a-a8f3-6dd3aa45da1b">

- The OAuth flow object is extended with 3 new fields for CIBA
- The authorizationUrl and tokenUrl fields  are updated to specify that they are also applicable for the CIBA flow

<img width="392" alt="image" src="https://github.com/OAI/OpenAPI-Specification/assets/77152136/6e977c13-1b9f-42b2-83ac-e61310baecfd">

- The OAuth flow object example is extended with a CIBA flow example
<img width="387" alt="image" src="https://github.com/OAI/OpenAPI-Specification/assets/77152136/7b440ac9-0369-4c2a-8a63-6dcf9b90273c">

This PR: 
Fixes #3587 

Branch and file used:
* 3.2.0 spec: v3.2.0-dev branch, versions/3.2.0.md

LasneF commented:
@shilpa-padgaonkar  : 

we should upgrade as well the schema 
https://github.com/OAI/OpenAPI-Specification/blob/v3.2.0-dev/schemas/v3.1/schema.yaml 

@handrews  this raise the point that the 3.2 should have its own schema version  (compatible but still its own version) 

to me the delivery mode should be an enum following the acceptable values
"backchannel_token_delivery_mode: REQUIRED. One of the following values: poll, ping, or push." 

about user code  , we may need to include it as part of the sample 
not sure if as well it should not be a boolean , not clear why a string array 

another point should it be called ciba , or should it be called backchannel as the OIDC spec mentionned those name as of parameter (ie it s just a legitimate question , to me ciba is fine as pointing the precise concept, not aware if there are other back channel technology leveraging those concept)

darrelmiller commented:
Could you help me understand why there is a need for a new OAuth2Flow when this information could be communicated via an OpenIDConnect configuration document considering it is an OIDC specific flow?

AxelNennker commented:
> Could you help me understand why there is a need for a new OAuth2Flow when this information could be communicated via an OpenIDConnect configuration document considering it is an OIDC specific flow?

Hi @darrelmiller although it sounds like "OpenID Connect Client-Initiated Backchannel Authentication Flow" is the same as OpenId Connect is is not.
Also an OpenId Provider can offer CIBA and Authorization Code flow and then the API consumer would not know if the API only allows CIBA (or is intended to be used with CIBA).

This idea of introducing ciba to OAI comes out of the Camara project where we discovered the need to specify this flow.
Many Camara APIs use CIBA. Many APIs from MobileConnect also use CIBA but the flow was communicated offline and not documented in the e.g. openapi.yaml. We found that the flow should be part of the Camara API specification, and here we are.
Hope this explains why we need this and why openIdConnectUrl is not enough.

shilpa-padgaonkar commented:
@LasneF : Thanks for your feedback.
- I have updated the schema files.
- fixed the type of ciba_user_code to boolean
- for ciba_delivery_modes, it is also possible that not just one, but rather 2 or all 3 of the delivery modes are selected.

darrelmiller commented on versions/3.2.0.md:
```diff
@@ -3300,11 +3301,14 @@ Configuration details for a supported OAuth Flow
 <a name="oauthFlowRefreshUrl"></a>refreshUrl | `string` | `oauth2` | The URL to be used for obtaining refresh tokens. This MUST be in the form of a URL. The OAuth2 standard requires the use of TLS.
 <a name="oauthFlowScopes"></a>scopes | Map[`string`, `string`] | `oauth2` | **REQUIRED**. The available scopes for the OAuth2 security scheme. A map between the scope name and a short description for it. The map MAY be empty.
+<a name="backchannel_token_delivery_modes_supported"></a>ciba_delivery_modes | Array[`string`] | `oauth2` (`"ciba"`) | **REQUIRED**. JSON array containing one or more of the following values: poll, ping, and push.
+<a name="backchannel_authentication_request_signing_alg_values"></a>ciba_signing_algs | Array[`string`] | `oauth2` (`"ciba"`) | **OPTIONAL**. JSON array containing a list of the JWS signing algorithms (alg values) supported by the OP (OpenIdConnect Provider) for signed authentication requests
```
```suggestion
<a name="backchannel_authentication_request_signing_alg_values"></a>ciba_signing_algs | Array[`string`] | `oauth2` (`"ciba"`) |  JSON array containing a list of the JWS signing algorithms (alg values) supported by the OP (OpenIdConnect Provider) for signed authentication requests
```

darrelmiller commented on versions/3.2.0.md:
```diff
@@ -3300,11 +3301,14 @@ Configuration details for a supported OAuth Flow
 <a name="oauthFlowScopes"></a>scopes | Map[`string`, `string`] | `oauth2` | **REQUIRED**. The available scopes for the OAuth2 security scheme. A map between the scope name and a short description for it. The map MAY be empty.
+<a name="backchannel_token_delivery_modes_supported"></a>ciba_delivery_modes | Array[`string`] | `oauth2` (`"ciba"`) | **REQUIRED**. JSON array containing one or more of the following values: poll, ping, and push.
+<a name="backchannel_authentication_request_signing_alg_values"></a>ciba_signing_algs | Array[`string`] | `oauth2` (`"ciba"`) | **OPTIONAL**. JSON array containing a list of the JWS signing algorithms (alg values) supported by the OP (OpenIdConnect Provider) for signed authentication requests
+<a name="backchannel_user_code_parameter_supported"></a>ciba_user_code | `boolean` | `oauth2` (`"ciba"`) | **OPTIONAL**. Boolean value specifying whether the OP supports the use of the user_code parameter, with true indicating support. If omitted, the default value is false.
```
```suggestion
<a name="backchannel_user_code_parameter_supported"></a>ciba_user_code | `boolean` | `oauth2` (`"ciba"`) |  Boolean value specifying whether the OP supports the use of the user_code parameter, with true indicating support. If omitted, the default value is false.
```

lornajane commented:
Sorry, I know we've had some back-and-forth on the schemas, but since this change won't go into 3.1, it would be better to omit the schema updates from this pull request (the 3.2 schemas haven't been created yet), we can't merge this pull request with these schema changes in.

darrelmiller commented on versions/3.2.0.md:
```diff
@@ -3300,11 +3301,14 @@ Configuration details for a supported OAuth Flow
+<a name="oauthFlowTokenUrl"></a>tokenUrl | `string` | `oauth2` (`"password"`, `"clientCredentials"`, `"authorizationCode"`, `"deviceAuthorization"`, `"ciba"`) | **REQUIRED**. The token URL to be used for this flow. This MUST be in the form of a URL. The OAuth2 standard requires the use of TLS.
 <a name="oauthFlowRefreshUrl"></a>refreshUrl | `string` | `oauth2` | The URL to be used for obtaining refresh tokens. This MUST be in the form of a URL. The OAuth2 standard requires the use of TLS.
 <a name="oauthFlowScopes"></a>scopes | Map[`string`, `string`] | `oauth2` | **REQUIRED**. The available scopes for the OAuth2 security scheme. A map between the scope name and a short description for it. The map MAY be empty.
+<a name="backchannel_token_delivery_modes_supported"></a>ciba_delivery_modes | Array[`string`] | `oauth2` (`"ciba"`) | **REQUIRED**. JSON array containing one or more of the following values: poll, ping, and push.
```
OpenAPI uses camelCasing for parameter values.  So although the CIBA specification uses snake case, for consistency we should require tooling to do the translation and maintain consistency in the OpenAPI description.

shilpa-padgaonkar replied:
```suggestion
<a name="backchannelTokenDeliveryModesSupported"></a>ciba_delivery_modes | Array[`string`] | `oauth2` (`"ciba"`) | **REQUIRED**. JSON array containing one or more of the following values: poll, ping, and push.
```

darrelmiller reviewed (changes requested):
Remove OPTIONAL keywords and change casing of parameters

handrews commented:
@darrelmiller 

> Remove OPTIONAL keywords and change casing of parameters

I presume you meant "Remove OPTIONAL ***from*** keywords..." rather than removing the keywords marked OPTIONAL

shilpa-padgaonkar commented on versions/3.2.0.md:
```diff
@@ -3300,11 +3301,14 @@ Configuration details for a supported OAuth Flow
 <a name="oauthFlowRefreshUrl"></a>refreshUrl | `string` | `oauth2` | The URL to be used for obtaining refresh tokens. This MUST be in the form of a URL. The OAuth2 standard requires the use of TLS.
 <a name="oauthFlowScopes"></a>scopes | Map[`string`, `string`] | `oauth2` | **REQUIRED**. The available scopes for the OAuth2 security scheme. A map between the scope name and a short description for it. The map MAY be empty.
+<a name="backchannelTokenDeliveryModesSupported"></a>ciba_delivery_modes | Array[`string`] | `oauth2` (`"ciba"`) | **REQUIRED**. JSON array containing one or more of the following values: poll, ping, and push.
+<a name="backchannel_authentication_request_signing_alg_values"></a>ciba_signing_algs | Array[`string`] | `oauth2` (`"ciba"`) |  JSON array containing a list of the JWS signing algorithms (alg values) supported by the OP (OpenIdConnect Provider) for signed authentication requests
```
```suggestion
<a name="backchannelAuthenticationRequestSigningAlgValues"></a>ciba_signing_algs | Array[`string`] | `oauth2` (`"ciba"`) |  JSON array containing a list of the JWS signing algorithms (alg values) supported by the OP (OpenIdConnect Provider) for signed authentication requests
```

shilpa-padgaonkar commented on versions/3.2.0.md:
```diff
@@ -3300,11 +3301,14 @@ Configuration details for a supported OAuth Flow
 <a name="oauthFlowScopes"></a>scopes | Map[`string`, `string`] | `oauth2` | **REQUIRED**. The available scopes for the OAuth2 security scheme. A map between the scope name and a short description for it. The map MAY be empty.
+<a name="backchannelTokenDeliveryModesSupported"></a>ciba_delivery_modes | Array[`string`] | `oauth2` (`"ciba"`) | **REQUIRED**. JSON array containing one or more of the following values: poll, ping, and push.
+<a name="backchannelAuthenticationRequestSigningAlgValues"></a>ciba_signing_algs | Array[`string`] | `oauth2` (`"ciba"`) |  JSON array containing a list of the JWS signing algorithms (alg values) supported by the OP (OpenIdConnect Provider) for signed authentication requests
+<a name="backchannel_user_code_parameter_supported"></a>ciba_user_code | `boolean` | `oauth2` (`"ciba"`) |  Boolean value specifying whether the OP supports the use of the user_code parameter, with true indicating support. If omitted, the default value is false.
```
```suggestion
<a name="backchannelUserCodeParameterSupported"></a>ciba_user_code | `boolean` | `oauth2` (`"ciba"`) |  Boolean value specifying whether the OP supports the use of the user_code parameter, with true indicating support. If omitted, the default value is false.
```

shilpa-padgaonkar commented on versions/3.2.0.md:
```diff
@@ -3300,11 +3301,14 @@ Configuration details for a supported OAuth Flow
+<a name="oauthFlowTokenUrl"></a>tokenUrl | `string` | `oauth2` (`"password"`, `"clientCredentials"`, `"authorizationCode"`, `"deviceAuthorization"`, `"ciba"`) | **REQUIRED**. The token URL to be used for this flow. This MUST be in the form of a URL. The OAuth2 standard requires the use of TLS.
 <a name="oauthFlowRefreshUrl"></a>refreshUrl | `string` | `oauth2` | The URL to be used for obtaining refresh tokens. This MUST be in the form of a URL. The OAuth2 standard requires the use of TLS.
 <a name="oauthFlowScopes"></a>scopes | Map[`string`, `string`] | `oauth2` | **REQUIRED**. The available scopes for the OAuth2 security scheme. A map between the scope name and a short description for it. The map MAY be empty.
+<a name="backchannelTokenDeliveryModesSupported"></a>ciba_delivery_modes | Array[`string`] | `oauth2` (`"ciba"`) | **REQUIRED**. JSON array containing one or more of the following values: poll, ping, and push.
```
```suggestion
<a name="backchannelTokenDeliveryModesSupported"></a>cibaDeliveryModes | Array[`string`] | `oauth2` (`"ciba"`) | **REQUIRED**. JSON array containing one or more of the following values: poll, ping, and push.
```

shilpa-padgaonkar commented on versions/3.2.0.md:
```diff
@@ -3300,11 +3301,14 @@ Configuration details for a supported OAuth Flow
 <a name="oauthFlowRefreshUrl"></a>refreshUrl | `string` | `oauth2` | The URL to be used for obtaining refresh tokens. This MUST be in the form of a URL. The OAuth2 standard requires the use of TLS.
 <a name="oauthFlowScopes"></a>scopes | Map[`string`, `string`] | `oauth2` | **REQUIRED**. The available scopes for the OAuth2 security scheme. A map between the scope name and a short description for it. The map MAY be empty.
+<a name="backchannelTokenDeliveryModesSupported"></a>cibaDeliveryModes | Array[`string`] | `oauth2` (`"ciba"`) | **REQUIRED**. JSON array containing one or more of the following values: poll, ping, and push.
+<a name="backchannelAuthenticationRequestSigningAlgValues"></a>ciba_signing_algs | Array[`string`] | `oauth2` (`"ciba"`) |  JSON array containing a list of the JWS signing algorithms (alg values) supported by the OP (OpenIdConnect Provider) for signed authentication requests
```
```suggestion
<a name="backchannelAuthenticationRequestSigningAlgValues"></a>cibaSigningAlgs | Array[`string`] | `oauth2` (`"ciba"`) |  JSON array containing a list of the JWS signing algorithms (alg values) supported by the OP (OpenIdConnect Provider) for signed authentication requests
```

shilpa-padgaonkar commented on versions/3.2.0.md:
```diff
@@ -3300,11 +3301,14 @@ Configuration details for a supported OAuth Flow
 <a name="oauthFlowScopes"></a>scopes | Map[`string`, `string`] | `oauth2` | **REQUIRED**. The available scopes for the OAuth2 security scheme. A map between the scope name and a short description for it. The map MAY be empty.
+<a name="backchannelTokenDeliveryModesSupported"></a>cibaDeliveryModes | Array[`string`] | `oauth2` (`"ciba"`) | **REQUIRED**. JSON array containing one or more of the following values: poll, ping, and push.
+<a name="backchannelAuthenticationRequestSigningAlgValues"></a>cibaSigningAlgs | Array[`string`] | `oauth2` (`"ciba"`) |  JSON array containing a list of the JWS signing algorithms (alg values) supported by the OP (OpenIdConnect Provider) for signed authentication requests
+<a name="backchannelUserCodeParameterSupported"></a>ciba_user_code | `boolean` | `oauth2` (`"ciba"`) |  Boolean value specifying whether the OP supports the use of the user_code parameter, with true indicating support. If omitted, the default value is false.
```
```suggestion
<a name="backchannelUserCodeParameterSupported"></a>cibaUserCode | `boolean` | `oauth2` (`"ciba"`) |  Boolean value specifying whether the OP supports the use of the user_code parameter, with true indicating support. If omitted, the default value is false.
```

shilpa-padgaonkar commented on versions/3.2.0.md:
```diff
@@ -3328,6 +3332,15 @@ This object MAY be extended with [Specification Extensions](#specificationExtens
+        "write:pets": "modify pets in your account",
+        "read:pets": "read your pets"
+      },
+      "ciba_delivery_modes": ["poll"]
```
```suggestion
      "cibaDeliveryModes": ["poll"]
```

shilpa-padgaonkar commented:
> Sorry, I know we've had some back-and-forth on the schemas, but since this change won't go into 3.1, it would be better to omit the schema updates from this pull request (the 3.2 schemas haven't been created yet), we can't merge this pull request with these schema changes in.

@lornajane Reverted schema changes

shilpa-padgaonkar commented:
@darrelmiller Done changes for camelCase. Kindly check.

ralfhandl commented on versions/3.2.0.md:
```diff
@@ -3300,11 +3301,14 @@ Configuration details for a supported OAuth Flow
+<a name="oauthFlowTokenUrl"></a>tokenUrl | `string` | `oauth2` (`"password"`, `"clientCredentials"`, `"authorizationCode"`, `"deviceAuthorization"`, `"ciba"`) | **REQUIRED**. The token URL to be used for this flow. This MUST be in the form of a URL. The OAuth2 standard requires the use of TLS.
 <a name="oauthFlowRefreshUrl"></a>refreshUrl | `string` | `oauth2` | The URL to be used for obtaining refresh tokens. This MUST be in the form of a URL. The OAuth2 standard requires the use of TLS.
 <a name="oauthFlowScopes"></a>scopes | Map[`string`, `string`] | `oauth2` | **REQUIRED**. The available scopes for the OAuth2 security scheme. A map between the scope name and a short description for it. The map MAY be empty.
+<a name="backchannelTokenDeliveryModesSupported"></a>cibaDeliveryModes | Array[`string`] | `oauth2` (`"ciba"`) | **REQUIRED**. JSON array containing one or more of the following values: poll, ping, and push.
```
```suggestion
<a name="backchannelTokenDeliveryModesSupported"></a>cibaDeliveryModes | Array[`string`] | `oauth2` (`"ciba"`) | **REQUIRED**. JSON array containing one or more of the following values: `poll`, `ping`, and `push`.
```

lornajane reviewed (approved):
Thank you for being patient with us and working through the amendments.
